Image featurization using ImageNet models

hsidky opened the pull request:
This pull request includes a plugin which uses pre-trained computer vision models (e.g. ResNet50, VGG19, etc...) to generate low-dimensional features for image collections. 

The current implementation is pretty basic. It assumes that each individual image can fit into memory since this plugin is likely going to be run on collections of smaller-sized images, rather than a single extremely large images. Support for reading tiles from the latter can be added later as needed. 

Downstream plugins can take the resulting CSV and use that for clustering and visualization tasks.

Nicholas-Schaub commented on polus-imagenet-model-featurization-plugin/src/main.py:
```diff
@@ -0,0 +1,142 @@
+    # Parse the arguments
+    args = parser.parse_args()
+
+    input_dir = args.inpDir
+    logger.info('inpDir = {}'.format(input_dir))
```
There is a quirk in WIPP where we need to check to see if there is an `images` subdirectory, and switch to it if it exists.

hsidky replied:
Will fix, thanks.

Nicholas-Schaub commented on utils/polus-bfio-util/bfio/bfio.py:
```diff
@@ -974,7 +974,7 @@ def __init__(self, file_path, image=None,
             self._metadata.image().Pixels.channel_count = self._xyzct['C']
             self._metadata.image().Pixels.DimensionOrder = bioformats.omexml.DO_XYZCT
-        elif image != None:
+        elif image is not None:
```
Usually this should be included in a different PR, but I don't think it matters.

hsidky replied:
Not sure how this got in. It must have been living in my repo and accidentally got merged back into my working branch.

Nicholas-Schaub commented on polus-imagenet-model-featurization-plugin/src/main.py:
```diff
@@ -0,0 +1,142 @@
+    image_files = [os.path.join(dirpath, file)
+                   for dirpath, _, files in os.walk(input_dir)
+                   for file in files 
+                   if os.path.isfile(os.path.join(dirpath, file)) and file.endswith('.ome.tif')]
```
Just a few thoughts.

There is no explicit need to use `bfio` for this particular task. However, if/when we switch over to using a different file format such as `ome.zarr`, using `bfio` would make the transition easier. Also, I suspect that `bfio` might increase the read speed of `ome.tif` files.

hsidky replied:
Thanks, I'll switch over to bfio.

Nicholas-Schaub reviewed (changes requested):
Just a minor change to the input and this is good to go.

hsidky commented:
@Nicholas-Schaub I addressed your comments. I profiled with bfio and found it to basically be identical to imageio. Since I'm not using the bfio container for tensorflow reasons I decided to stick with imageio. Let me know if you request any more changes.

hsidky commented:
Switched to bfio 2 per our discussion using python backend.
